cpu/stm32f4/periph/spi.c

```diff
@@ -102,6 +111,9 @@ int spi_init_master(spi_t dev, spi_conf_t conf, spi_speed_t speed)
             return -2;
     }
 
+    /* initialize lock */
+    mutex_init(&spi_config[dev].lock);
```

Comment:
@gebart  basically the mutex initialization [resets](https://github.com/RIOT-OS/RIOT/blob/master/core/include/mutex.h#L61) the mutex. 
In this case initializing, or resetting, a specific `dev` into SPI master mode must reset the mutex too as it could be persist locked from a previous state and produce a deadlock.  

> On the other hand, I would prefer that all calls to spi_init would be moved to each platform's board.c instead of...

Another proposition to prevent reinitialization of the mutex and a running device would be to store the initialized state in the `spi_state_t` of the `dev` and to skip `spi_init_master()` for the `dev` if its done already.
Resetting the stored state could be done in [`spi_poweroff()`](https://github.com/haukepetersen/RIOT/blob/opt_spi_threadsafe/cpu/stm32f4/periph/spi.c#L429) for the specific `dev`.

+1 to move the mutex to the top and acquire it before doing initializations